forc-plugins/forc-doc/src/doc.rs

```diff
@@ -49,17 +49,35 @@ impl Document {
         }
     }
     fn preview_opt(&self) -> Option<String> {
-        match &self.raw_attributes {
-            Some(description) => {
-                if description.len() > 200 {
-                    let (first, _) = description.split_at(75);
-                    Some(format!("{}...", first.trim_end()))
-                } else {
-                    Some(description.to_string())
+        const MAX_PREVIEW_CHARS: usize = 100;
+        const CLOSING_PARAGRAPH_TAG: &str = "</p>";
+
+        self.raw_attributes.as_ref().map(|description| {
+            let preview = split_at_markdown_header(description);
+            if preview.chars().count() > MAX_PREVIEW_CHARS
+                && preview.contains(CLOSING_PARAGRAPH_TAG)
+            {
+                match preview.find(CLOSING_PARAGRAPH_TAG) {
+                    Some(index) => {
+                        // We add 1 here to get the index of the char after the closing tag.
+                        // This ensures we retain the closing tag and don't break the html.
+                        let (preview, _) =
+                            preview.split_at(index + CLOSING_PARAGRAPH_TAG.chars().count() + 1);
```

Comment:
Careful: `split_at` expects a byte index, not a character index: https://doc.rust-lang.org/std/primitive.str.html#method.split_at. Here, `index` is a byte index produced by `find`, but the right-hand-side of the add is in `chars`.

In general, working with `str` requires being careful about distinguishing between bytes and chars, so it's always worth checking what methods like `split_at` or `find` expect. It gets even more complicated when working with graphemes and emoji :sweat:

Reply:
Ah thanks, I'll read more into this tomorrow morning!

Reply:
Okay, so I took a look just now. I've altered it to check in `chars` for the conditional statement but then use the `bytes` for splitting since it's checking for a specific index. I can't check how it looks just yet because the `forc` building issue, but this is essentially what I had before, save that the initial conditional statement now checks for the length in `chars`.

Reply:
Just checked with `--locked` and looks like it's functioning as intended